File: src/main/java/sage/entity/Tag.java
package sage.entity;

import java.util.Collection;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;

import sage.domain.IdCommons;

/**
 * This should be a permanently cached entity-type
 */
@Entity(name = "Tag")
public class Tag {
  public static final long ROOT_ID = 1;
  public static final String ROOT_NAME = "无";

  private Long id;
  private String name;
  private boolean isCore;
  private Set<Tag> children = new HashSet<>();
  private Tag parent;

  private Collection<Tweet> tweets;
  private Collection<Blog> blogs;
  private Collection<Follow> follows;

  public Tag() {
  }

  public Tag(String name, Tag parent) {
    this.name = name;
    this.parent = parent;
  }

  @Id
  @GeneratedValue
  public Long getId() {
    return id;
  }

  public void setId(long id) {
    this.id = id;
  }

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }
  
  public boolean isCore() {
    return isCore;
  }
  
  public void setCore(boolean isCore) {
    this.isCore = isCore;
  }

  @OneToMany(mappedBy = "parent")
  public Set<Tag> getChildren() {
    return children;
  }

  public void setChildren(Set<Tag> children) {
    this.children = children;
  }

  @ManyToOne
  public Tag getParent() {
    return parent;
  }

  public void setParent(Tag parent) {
    this.parent = parent;
  }

  @ManyToMany(mappedBy = "tags")
  public Collection<Tweet> getTweets() {
    return tweets;
  }

  public void setTweets(Collection<Tweet> tweets) {
    this.tweets = tweets;
  }

  @ManyToMany(mappedBy = "tags")
  public Collection<Blog> getBlogs() {
    return blogs;
  }

  public void setBlogs(Collection<Blog> blogs) {
    this.blogs = blogs;
  }

  @ManyToMany(mappedBy = "tags")
  public Collection<Follow> getFollows() {
    return follows;
  }

  public void setFollows(Collection<Follow> follows) {
    this.follows = follows;
  }

  /**
   * @return a chain from itself to ancestors, excluding root; is empty for root
   */
  public List<Tag> chainUp() {
    List<Tag> chain = new LinkedList<>();
    if (getId() == ROOT_ID)
      return chain;

    Tag current = this;
    while (current.getId() != ROOT_ID) {
      chain.add(current);
      current = current.getParent();
    }
    return chain;
  }

  /**
   * @return all of its descendant tags
   */
  public Set<Tag> descendants() {
    Set<Tag> descanants = new HashSet<>();
    for (Tag child : getChildren()) {
      descanants.add(child);
      descanants.addAll(child.descendants());
    }
    return descanants;
  }

  @Override
  public String toString() {
    return name;
  }

  @Override
  public int hashCode() {
    return IdCommons.hashCode(getId());
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    
    Tag other = (Tag) obj;
    return IdCommons.equal(getId(), other.getId());
  }
}

File: src/main/kotlin/sage/entity/Tag.kt
package sage.entity

import sage.domain.commons.DomainException
import sage.transfer.TagCard
import sage.transfer.TagLabel
import sage.transfer.TagNode
import java.util.*
import javax.persistence.*

@Entity
class Tag(
    var name: String,
    @ManyToOne
    var parent: Tag?,
    var isCore: Boolean,
    @Column(columnDefinition = "TEXT")
    var intro: String = "",
    val creatorId: Long) : AutoModel() {

  @OneToMany(mappedBy = "parent")
  var children: Set<Tag> = HashSet()

  /**
   * @return a chain from itself to ancestors, excluding root; is empty for root
   */
  fun chainUp(): List<Tag> = listOf(this) + ancestorsUp()

  fun ancestorsUp(): List<Tag> {
    val ancestors = LinkedList<Tag>()
    if (id == ROOT_ID) {
      return ancestors
    }
    var current: Tag = this.parent!!
    while (current.id != ROOT_ID) {
      ancestors.add(current)
      current = current.parent!!
    }
    return ancestors
  }

  @Suppress("unused")
  fun chainDown(): List<Tag> = chainUp().asReversed()

  /**
   * @return all of its descendant tags
   */
  fun descendants(): Set<Tag> {
    val descendants = HashSet<Tag>()
    for (child in children) {
      descendants.add(child)
      descendants.addAll(child.descendants())
    }
    return descendants
  }

  fun getQueryTags(): Set<Tag> = descendants() + this

  @PrePersist @PreUpdate
  fun nameMustBeUnique() {
    siblingsCannotHaveThatName(name)
  }

  fun siblingsCannotHaveThatName(name: String) {
    parent?.let { parent ->
      if (parent.descendants().find { it.id != id && it.name == name } != null) {
        throw DomainException("标签\"${parent.name}\"之下有同名标签: \"$name\"")
      }
    }
  }

  fun toTagNode(): TagNode {
    val t = TagNode()
    t.id = id
    t.name = name
    t.isCore = isCore
    t.children = children
        .filter { it.isCore }
        .map { it.toTagNode() }
    return t
  }

  fun toTagCard(): TagCard {
    val t = TagCard()
    t.id = id
    t.name = name
    t.intro = intro
    t.isCore = isCore

    for (node in chainUp()) {
      t.chainUp.add(node.toTagLabel())
    }

    for (child in children) {
      t.children.add(child.toTagLabel())
    }

    t.chainStr = produceChainStr()

    return t
  }

  fun toTagLabel(): TagLabel {
    val t = TagLabel()
    t.id = id
    t.name = name
    t.isCore = isCore
    t.chainStr = produceChainStr()
    return t
  }

  private fun produceChainStr() = chainUp().asReversed().map(Tag::name).joinToString("->")

  companion object : BaseFind<Long, Tag>(Tag::class) {
    val ROOT_ID: Long = 1
    val ROOT_NAME = "无"

    fun multiGet(ids: Collection<Long>) = ids.mapNotNull { byId(it) }.toMutableSet()

    fun getQueryTags(tags: Collection<Tag>): Set<Tag> {
      val queryTags = HashSet<Tag>()
      for (node in tags) {
        queryTags.add(node)
        queryTags.addAll(node.descendants())
      }
      return queryTags
    }
  }
}